File: messageformat/mf1_test.go
package messageformat

import (
	"errors"
	"fmt"
	"strings"
	"testing"
)

// Behaviour of ParseMF1 beyond the shared fixture (testdata/glossa/mf1-to-mf2.json).

func TestParseMF1Errors(t *testing.T) {
	tests := []struct {
		name   string
		src    string
		locale string
		want   ErrorCode
	}{
		{"invalid locale", "Hallo", "not a locale!", CodeInvalidLocale},
		{"positional argument is not an MF2 name", "{0} Einträge", "de", CodeMF1Unsupported},
		{"selector conflict", "{x, select, a {A} other {B}} {x, selectordinal, one {1} other {n}}", "en", CodeMF1Unsupported},
		{"nested argument in a style", "{n, number, {x}}", "en", CodeMF1Unsupported},
		{
			name:   "too many variants",
			src:    manySelects(5), // 8^5 variants
			locale: "en",
			want:   CodeMF1Unsupported,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := ParseMF1(tt.src, tt.locale)
			if !errors.Is(err, &Error{Code: tt.want}) {
				t.Errorf("ParseMF1(%q) = %v, want %s", tt.src, err, tt.want)
			}
		})
	}
}

// manySelects returns n sibling selects with eight keys each.
func manySelects(n int) string {
	var b strings.Builder
	for i := range n {
		fmt.Fprintf(&b, "{v%d, select, a {a} b {b} c {c} d {d} e {e} f {f} g {g} other {x}}", i)
	}
	return b.String()
}

func TestParseMF1ToMF2Syntax(t *testing.T) {
	tests := []struct {
		name string
		src  string
		want string
	}{
		{
			name: "number pattern inside a plural keeps its # characters",
			src:  "{n, plural, other {{n, number, #,##0.0} Punkte}}",
			want: ".input {$n :number}\n.match $n\n* {{{$n :mf1:number mf1:argStyle=|#,##0.0| @mf1:argStyle=|#,##0.0| @mf1:argType=number} Punkte}}",
		},
		{
			name: "offset local avoids a taken name",
			src:  "{n, plural, offset:1 one {{n_minus_1} #} other {#}}",
			want: ".input {$n :number}\n.local $n_minus_1_2 = {$n :offset subtract=1}\n.match $n_minus_1_2\none {{{$n_minus_1} {$n_minus_1_2}}}\n* {{{$n_minus_1_2}}}",
		},
		{
			name: "exact key with leading zero",
			src:  "{n, plural, =01 {eins} other {viele}}",
			want: ".input {$n :number}\n.match $n\n1 {{eins}}\n* {{viele}}",
		},
		{
			name: "unknown currency skeleton falls back",
			src:  "{p, number, ::currency/eur}",
			want: "{$p :mf1:number mf1:argStyle=|::currency/eur| @mf1:argStyle=|::currency/eur| @mf1:argType=number}",
		},
		{
			name: "custom time pattern falls back",
			src:  "{t, time, HH:mm}",
			want: "{$t :mf1:time mf1:argStyle=|HH:mm| @mf1:argStyle=|HH:mm| @mf1:argType=time}",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			msg, err := ParseMF1(tt.src, "de")
			if err != nil {
				t.Fatalf("ParseMF1: %v", err)
			}
			got, err := Stringify(msg)
			if err != nil {
				t.Fatalf("Stringify: %v", err)
			}
			want, err := ParseMF2(tt.want)
			if err != nil {
				t.Fatalf("ParseMF2(want): %v", err)
			}
			wantSrc, _ := Stringify(want)
			if got != wantSrc {
				t.Errorf("ParseMF1(%q) =\n%s\nwant\n%s", tt.src, got, wantSrc)
			}
		})
	}
}

func TestParseMF1VariantsDoNotShareState(t *testing.T) {
	msg, err := ParseMF1("{g, select, a {x} other {y}} {total, number, ::currency/EUR}", "de")
	if err != nil {
		t.Fatal(err)
	}
	first := msg.Variants[0].Value[1].(Expression)
	first.Function.Options["currency"] = Literal{Value: "USD"}
	second := msg.Variants[1].Value[1].(Expression)
	if got := second.Function.Options["currency"]; got != (Literal{Value: "EUR"}) {
		t.Errorf("mutating one variant changed another: %v", got)
	}
}

File: messageformat/js/src/mf1.test.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import { MF1SyntaxError, parseMF1 } from "./mf1.js";
import { format, parseMF2, stringify } from "./mf2.js";
import { isMessage } from "./schema.js";
import { testdataDir } from "./testing/index.js";

/** Convert, then show as MF2 syntax without the round-trip attributes. */
const mf2 = (src: string, locale = "en") =>
  stringify(parseMF1(src, locale))
    .replace(/ @mf1:arg(Type|Style)=(\|[^|]*\||[^\s}]+)/g, "")
    .replace(/\n/g, " ");

describe("parseMF1 maps ICU MessageFormat 1 onto standard MF2 functions", () => {
  it.each([
    ["Hello {name}", "Hello {$name}"],
    ["It's '{literal}'", "It's \\{literal\\}"],
    ["{n, number}", "{$n :number}"],
    ["{n, number, integer}", "{$n :integer}"],
    ["{n, number, percent}", "{$n :percent}"],
    [
      "{n, number, ::percent scale/100 .0}",
      "{$n :percent minimumFractionDigits=1 maximumFractionDigits=1}",
    ],
    ["{n, number, ::currency/EUR}", "{$n :currency currency=EUR}"],
    ["{n, number, ::.00}", "{$n :number minimumFractionDigits=2 maximumFractionDigits=2}"],
    ["{n, number, ::measure-unit/length-kilometer}", "{$n :unit unit=kilometer}"],
    ["{d, date}", "{$d :date}"],
    ["{d, date, short}", "{$d :date length=short}"],
    ["{d, date, long}", "{$d :date length=long}"],
    ["{d, date, full}", "{$d :date fields=year-month-day-weekday length=long}"],
    ["{t, time}", "{$t :time precision=second}"],
    ["{t, time, short}", "{$t :time precision=minute}"],
    ["{t, time, full}", "{$t :time precision=second timeZoneStyle=long}"],
    [
      "{g, select, female {She} male {He} other {They}}",
      ".input {$g :string} .match $g female {{She}} male {{He}} * {{They}}",
    ],
    [
      "{c, plural, =0 {none} one {# item} other {# items}}",
      ".input {$c :number} .match $c 0 {{none}} one {{{$c} item}} * {{{$c} items}}",
    ],
    [
      "{c, selectordinal, one {#st} two {#nd} few {#rd} other {#th}}",
      ".input {$c :number select=ordinal} .match $c one {{{$c}st}} two {{{$c}nd}} few {{{$c}rd}} * {{{$c}th}}",
    ],
  ])("%s → %s", (src, exp) => {
    expect(mf2(src)).toBe(exp);
  });

  it("lifts nested selectors into one .match with the catch-all last", () => {
    const src =
      "{g, select, female {{c, plural, one {She has # cat} other {She has # cats}}} " +
      "other {{c, plural, one {They have # cat} other {They have # cats}}}}";
    expect(mf2(src)).toBe(
      ".input {$g :string} .input {$c :number} .match $g $c " +
        "female one {{She has {$c} cat}} female * {{She has {$c} cats}} " +
        "* one {{They have {$c} cat}} * * {{They have {$c} cats}}",
    );
  });

  it("maps a plural offset onto :offset, keeping exact matches on the raw value", () => {
    const src =
      "{c, plural, offset:1 =0 {Nobody} =1 {{host}} one {{host} and # other} " +
      "other {{host} and # others}}";
    expect(mf2(src)).toBe(
      ".input {$c :number} .local $c_offset = {$c :offset subtract=1} .match $c $c_offset " +
        "1 * {{{$host}}} 0 * {{Nobody}} * one {{{$host} and {$c_offset} other}} " +
        "* * {{{$host} and {$c_offset} others}}",
    );
    const msg = parseMF1(src, "en");
    const fmt = (c: number) => format(msg, "en", { c, host: "Ada" }, { bidiIsolation: "none" });
    expect([0, 1, 2, 3].map(fmt)).toEqual(["Nobody", "Ada", "Ada and 1 other", "Ada and 2 others"]);
  });

  it("avoids a name clash for the offset variable", () => {
    const out = mf2("{c, plural, offset:1 one {# {c_offset}} other {#}}");
    expect(out).toContain(".local $c_offset_ = {$c :offset subtract=1}");
    expect(out).toContain("{{{$c_offset_} {$c_offset}}}");
  });

  it("falls back to mf1: functions only where MF2 has no equivalent", () => {
    expect(mf2("{n, number, currency}")).toBe("{$n :mf1:currency}");
    expect(mf2("{n, number, ::compact-short}")).toMatch(/^\{\$n :mf1:number /);
    expect(mf2("{d, date, ::yMMMd}")).toMatch(/^\{\$d :mf1:date /);
    expect(mf2("{n, spellout}")).toBe("{$n :mf1:spellout}");
    expect(mf2("{d, duration}")).toBe("{$d :mf1:duration}");
  });

  it("keeps the MF1 argument type and style as attributes for export", () => {
    const msg = parseMF1("{n, number, percent}", "en");
    expect(msg).toMatchObject({
      pattern: [
        {
          attributes: {
            "mf1:argType": { type: "literal", value: "number" },
            "mf1:argStyle": { type: "literal", value: "percent" },
          },
        },
      ],
    });
  });

  it("produces plain JSON matching the data model schema", () => {
    const msg = parseMF1("{n, number, ::.00} {c, plural, offset:2 =0 {a} other {#}}", "de");
    expect(isMessage(msg)).toBe(true);
    expect(JSON.parse(JSON.stringify(msg))).toEqual(msg);
    expect(parseMF2(stringify(msg))).toEqual(msg);
  });

  it("validates plural keys against the locale", () => {
    expect(() => parseMF1("{n, plural, one {a} few {b} other {c}}", "en")).toThrow(MF1SyntaxError);
    expect(() => parseMF1("{n, plural, one {a} few {b} many {c} other {d}}", "pl")).not.toThrow();
  });

  it("reports syntax errors as MF1SyntaxError", () => {
    expect(() => parseMF1("{n, plural, one {a}", "en")).toThrow(MF1SyntaxError);
  });

  it("formats converted German messages", () => {
    const msg = parseMF1(
      "{count, plural, =0 {Keine Dateien} one {Eine Datei} other {# Dateien}} ({size, number, ::.0} MB)",
      "de",
    );
    const out = format(msg, "de", { count: 1234, size: 3.25 }, { bidiIsolation: "none" });
    expect(out).toBe("1.234 Dateien (3,3 MB)");
  });
});

/**
 * Shared fixture owned by the Go converter: MF1 source → canonical MF2 data
 * model. Both implementations must produce identical data models.
 * Shape (mirrors the Unicode suite): { "tests": [{ "src", "locale"?, "exp" }] }.
 */
const sharedFixture = join(testdataDir, "glossa/mf1-to-mf2.json");

describe.skipIf(!existsSync(sharedFixture))("shared fixture glossa/mf1-to-mf2.json", () => {
  interface Case {
    description?: string;
    src: string;
    locale?: string;
    exp: unknown;
  }
  const load = (): Case[] => {
    const file = JSON.parse(readFileSync(sharedFixture, "utf8")) as {
      defaultTestProperties?: Partial<Case>;
      tests: Case[];
    };
    return file.tests.map((t) => ({ ...file.defaultTestProperties, ...t }));
  };
  it.each(existsSync(sharedFixture) ? load().map((c) => [c.description ?? c.src, c] as const) : [])(
    "%s",
    (_name, c) => {
      expect(parseMF1(c.src, c.locale ?? "en")).toEqual(c.exp);
    },
  );
});

if (!existsSync(sharedFixture)) {
  console.info(`skipping shared MF1 fixture: ${sharedFixture} not present`);
}
